internal/controller/api_to_config.go

```diff
@@ -96,12 +97,55 @@ func neighborToFRR(n v1beta1.Neighbor, ipv4Prefixes, ipv6Prefixes []string) (*fr
 		case ipfamily.IPv6:
 			res.HasV6Advertisements = true
 		}
-		res.Advertisements = append(res.Advertisements, &frr.AdvertisementConfig{Prefix: p, IPFamily: family})
+
+		// TODO: check that the prefix matches the passed IPv4/IPv6 prefixes
+		advs[p] = &frr.AdvertisementConfig{Prefix: p, IPFamily: family}
+	}
+
+	prefixToCommunities := map[string]sets.Set[string]{}
+	for _, pfxs := range n.ToAdvertise.PrefixesWithCommunity {
+		// TODO: add community format verification
+		for _, p := range pfxs.Prefixes {
+			_, ok := advs[p]
+			if !ok {
+				return nil, fmt.Errorf("prefix %s with community %s not in allowed list for neighbor %s", p, pfxs.Community, n.Address)
+			}
+			_, ok = prefixToCommunities[p]
+			if !ok {
+				prefixToCommunities[p] = sets.New(pfxs.Community)
+				continue
+			}
+
+			prefixToCommunities[p].Insert(pfxs.Community)
+		}
+	}
+
+	for p, c := range prefixToCommunities {
+		adv, ok := advs[p]
+		if !ok { // shouldn't happen as we check in previous loop, just in case
+			return nil, fmt.Errorf("unexpected err - no community prefix matching %s", p)
+		}
+		adv.Communities = sets.List(c)
```

Comment:
do we have a guaranteed order here?

Reply:
yes, the List func is explicitly for returning the contents sorted  ( as opposed to the c.UnsortedList() method)